### ISSUE TITLE
About the slicedis class

### ISSUE THREAD

**Issue by will876** [swhan0329/erp_inpainting#9]
Hello, I have a question. In the SliceDis class, I found that the final output shape of the model is [1, 6]. Does it identify the entire cube panoramic image dimension at one time and the dimension is 6? But what you write in your paper is to identify one face (batchsize, 1). Is there any problem with my understanding? Please enlighten me. Thank you!

**Comment by swhan0329** [comment]
@will876 
First of all, thank you for your interest in my code and thesis, it may have been confusing because of the many py's in the model folder, but I used the model structure in PIUnet2.py.

The corresponding SliceDis class can be found via the link below. 
https://github.com/swhan0329/panorama_image_inpainting/blob/13823f0dfeb9628bb20ba3c2154e5cbdf8918fa7/model/PIUnet2.py#L258

If you look at this code, you can see that the output is 1.
https://github.com/swhan0329/panorama_image_inpainting/blob/13823f0dfeb9628bb20ba3c2154e5cbdf8918fa7/train.py#L101

**Comment by will876** [comment]
Thank you very much for your answer. I am a beginner and my understanding of knowledge is relatively shallow. I still don’t quite understand z = x[:, 6 * k:6 * k + 6, :, :]. Didn’t z take 6 of the channels?
In addition, I would like to ask, does slicedis loop 6 times during the for loop, and then identify one face of the cube each time?

**Comment by swhan0329** [comment]
@will876 
Converting an image in Equirectangular format to a cube map splits it into 6 images. sliceDis determines whether the 6 images are true or false one by one, while WholeDis combines all six images at once to determine whether they are true or false.

SliceDis takes a cube map image and mask as input.
Since it's a cube map image, the input comes in with a total of 6 faces combined, and it receives the value of 1 face of each cube as 6*k:6*k+6.

**Comment by will876** [comment]
Thank you for your patient explanation. Does your code have two generators and three discriminators? If I want to make some changes based on yours, should I make changes to the generator and discriminator? I think the panoramic code is quite complicated, and the parts that are too difficult will not be modified. In addition, there are really few codes for panoramic repair. I only found your code, haha.

**Comment by swhan0329** [comment]
@will876 
I suggest you take a good look at the paper, it has images to help you understand it.
Also, I would recommend modifying both the Generator and the Discriminator. 

In particular, I think the model would perform better if the Discriminator added correlation information for each side of the cube.

**Comment by will876** [comment]
I will read your paper carefully. Thanks for your advice! I will give it a try as you said! If the loss of the discriminator is negative during the loop, does it prove that the current generator is stronger than the discriminator, so the discriminator can be modified to be better.

**Comment by will876** [comment]
@swhan0329 Sorry to bother you again. What does netfaceG do? Is it to generate the entire cube panoramic image? Does the entire process receive a cube panoramic image of .json type, then repair each individual face (CubeGenerator, WholeDis, SliceDis), then repair the entire 6 faces (FaceGenerator, FaceDis), and perform back propagation to get Loss the value, convert the cube into an equirectangular panoramic image to view the results, and the training is completed.

**Comment by swhan0329** [comment]
@will876 
The proposed network inputs a 360-degree equirectangular format panoramic image converts it into a cube map format, which has relatively little distortion and uses it as a training network. Since the cube map format is used, the correlation of the six sides of the cube map should be considered. Therefore, all faces of the cube map are used as input for the whole discriminative network, and each face of the cube map is used as input for the slice discriminative network to determine the authenticity of the generated image. The proposed network performed qualitatively better than existing single-image inpainting algorithms and baseline algorithms.

1. equirectangular format panoramic image -> cubemap format images(6 images)
2. cubemap format images -> generate using netFaceG -> Generated images
3. Whole generated images(6 images) -> WholeDis
4. Each generated images -> SliceDis

**Comment by will876** [comment]
Thank you very much and wish you good luck in your work!

**Comment by will876** [comment]
@swhan0329 Sorry to bother you again. If I want to change the generator and discriminator of the model, I directly take the generator and discriminator from another article about gan image repair, keep the input and output unchanged, and change the number of input and output channels and data set Is it ok to make some simple modifications like processing, or do I need to make many modifications according to my task. Or I copy the layers that exist in other discriminators and replace the layers in my discriminator.

**Comment by swhan0329** [comment]
@will876 I think it depends on what model you use. If the code is written so that the values ​​can vary depending on the input, you don't have to worry too much, but if not, you will need to modify the model layer to match the input size.

**Comment by will876** [comment]
@swhan0329       
 Hello, I want to ask two questions.
       First, I divided the 360SP data set into a training set and a test set. Then after the train is completed, test is performed by using test to read the .pth training model file saved in logs_s256. But why is the psnr output after each batch of my batch single digits instead of more than 30?
       The second question is, after training and testing, what code is used to obtain the parameters in the Table of the quantitative results? Is this code for measuring SSIM and PSNR included in your project? Thank you!

**Comment by swhan0329** [comment]
@will876 
1. Maybe ur training is failed. I wrote just used data set, it will be failed. Because in the dataset scenes are mixed such as city and nature. You need to divide it also. 
2. My code doesn't have evaluation code of SSIM and PSNR. You need to use other library.

**Comment by will876** [comment]
@swhan0329 
 I saw that my training result image was repaired well, but I didn’t expect that the psnr value would be so low.
Is the reason why my psnr value was too low during the test because I did not distinguish between cities and nature during training? If I did, would the psnr value be normal?
If I want to distinguish the data sets, I can only pick them out one by one. Is there no good way？

**Comment by swhan0329** [comment]
@will876 I think you need to analyse ur data and results. That area is not my area.
Dividing the dataset, you can see in your barely eyes or using green pixel values you can make classifier.

**Comment by will876** [comment]
@swhan0329 
Can I use your test model to inpainting my test set, and then use code found elsewhere to measure psnr?
But why do I think the effect of image inpainting is passable, but the printed PSNR is very low during the test?
If normally speaking, if the psnr I print for each test is single digit, will the psnr that needs to be placed in the paper table eventually also be single digit?

**Comment by swhan0329** [comment]
@will876  There is no weight for my model. At that time, people don't share the pretrained weight.

And in the case of inpainting tasks, the PSNR may be low, so it is not irritating to the human eye, but there may be differences at the pixel level. 

I think there are a lot of questions you have to think about, search for, and find answers for yourself. 

I hope that you will build a deeper understanding of this field and ask again, not just because it is a problem with my model.

I think I have given you as much help as I can, so if you have any other questions other than the concept or result analysis, please leave them.

**Comment by will876** [comment]
OK, thank you very much.
